### PR TITLE
[core] Refactor HeadlessFrontend/Backend: GL separation and factory

### PR DESCRIPTION
**Pull request by astojilj**
Refactor out HeadlessFrontend and HeadlessBackend gl independent code to gfx.
Define gl::HeadlessBackend as subclass, instantiated by gfx::HeadlessBackend static factory method.
Overriding the factory enabled.
GL dependent tests are still using gl::HeadlessBackend directly (not through gfx).

### CODE REVIEW

File: platform/default/include/mbgl/gfx/headless_backend.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace mbgl {
+namespace gfx {
+
+class HeadlessBackend : public gfx::Renderable {
+public:
+    // Factory.
+    static std::unique_ptr<HeadlessBackend> makeBackend(Size = { 256, 256 }, gfx::ContextMode = gfx::ContextMode::Unique);
+    
+    HeadlessBackend(Size);
```
Comment (pozdnyakov) [inline platform/default/include/mbgl/gfx/headless_backend.hpp]: protected?

File: platform/default/include/mbgl/gfx/headless_backend.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#include <mbgl/gfx/renderable.hpp>
+#include <mbgl/gfx/renderer_backend.hpp>
+#include <mbgl/util/image.hpp>
+
+#include <memory>
+
+namespace mbgl {
+namespace gfx {
+
+class HeadlessBackend : public gfx::Renderable {
```
Comment (pozdnyakov) [inline platform/default/include/mbgl/gfx/headless_backend.hpp]: nit: a short class description would be nice

File: platform/default/src/mbgl/gfx/headless_backend.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#include <mbgl/gfx/headless_backend.hpp>
+
+namespace mbgl {
+namespace gfx {
+
+HeadlessBackend::HeadlessBackend(Size size_)
```
Comment (pozdnyakov) [inline platform/default/src/mbgl/gfx/headless_backend.cpp]: do we need some asserts on the given `size_`?

File: platform/default/src/mbgl/gfx/headless_backend.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    : mbgl::gfx::Renderable(size_, nullptr) {
+}
+
+void HeadlessBackend::setSize(Size size_) {
```
Comment (pozdnyakov) [inline platform/default/src/mbgl/gfx/headless_backend.cpp]: ditto

File: platform/default/include/mbgl/gfx/headless_backend.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+public:
+    // Factory.
+    static std::unique_ptr<HeadlessBackend> makeBackend(Size = { 256, 256 }, gfx::ContextMode = gfx::ContextMode::Unique);
```
Comment (pozdnyakov) [inline platform/default/include/mbgl/gfx/headless_backend.hpp]: maybe just `create()`?

Comment (astojilj) [reply platform/default/include/mbgl/gfx/headless_backend.hpp]: Thanks. I put just make().

File: platform/default/src/mbgl/gfx/headless_backend.cpp (hunk 3 above)
Comment (astojilj) [reply platform/default/src/mbgl/gfx/headless_backend.cpp]: No need, lower layers of API allow 0.

File: platform/default/include/mbgl/gfx/headless_backend.hpp (hunk 2 above)
Comment (astojilj) [reply platform/default/include/mbgl/gfx/headless_backend.hpp]: Done.

File: platform/default/include/mbgl/gfx/headless_backend.hpp (hunk 1 above)
Comment (astojilj) [reply platform/default/include/mbgl/gfx/headless_backend.hpp]: Done. Although it is abstract class with factory method, no need to promote super class protected.

File: platform/default/src/mbgl/gl/headless_backend.cpp [hunk 6]
```diff
@@ -72,22 +72,27 @@ gfx::Renderable& HeadlessBackend::getDefaultRenderable() {
...
+}
 
 } // namespace gl
+
+#ifndef OVERRIDE_HEADLESS_BACKEND_FACTORY
+// Default factory implementation.
+std::unique_ptr<gfx::HeadlessBackend> gfx::HeadlessBackend::make(Size size, gfx::ContextMode contextMode) {
+    return std::make_unique<gl::HeadlessBackend>(size, contextMode);
+}
+#endif
```
[the remark this answers is not in the document]
Comment (astojilj) [reply platform/default/src/mbgl/gl/headless_backend.cpp]: As agreed offline, moved to follow up patch.

**Pull request merged**